Integrations/integration-Active_Directory_Query.yml

```diff
@@ -0,0 +1,1205 @@
+description: Active Direcory Query integration enables accessing and managing Active Directory objects (users,contacts and computers).
+configuration:
+- display: The server IP address
+  name: server_ip
+  defaultvalue: ""
+  type: 0
+  required: true
+- display: Port
+  name: port
+  defaultvalue: ""
+  type: 0
+  required: false
+- display: Credentials
+  name: credentials
+  defaultvalue: ""
+  type: 9
+  required: true
+- display: Base DN
+  name: base_dn
+  defaultvalue: ""
+  type: 0
+  required: true
+- display: Page size
+  name: page_size
+  defaultvalue: "500"
+  type: 0
+  required: true
+- display: Secure Connection
+  name: secure_connection
+  defaultvalue: None
+  type: 15
+  required: true
+  options:
+  - None
+  - SSL
+  - TLS
+script:
+  script: |+
+    from ldap3 import Server, Connection, NTLM, SUBTREE, ALL, ALL_ATTRIBUTES
+    from ldap3.extend import microsoft
+    from datetime import datetime, timedelta
+
+
+    ''' INSTANCE CONFIGURATION '''
+    SERVER_IP = demisto.params().get('server_ip')
+    PORT = demisto.params().get('port')
+    USERNAME = demisto.params().get('credentials')['identifier']
+    PASSWORD = demisto.params().get('credentials')['password']
+    DEFAULT_BASE_DN = demisto.params().get('base_dn')
+    SECURE_CONNECTION = demisto.params().get('secure_connection')
+    DEFAULT_PAGE_SIZE = int(demisto.params().get('page_size'))
+
+    if PORT:
+        PORT = int(PORT)
+
+    ''' GLOBAL VARS '''
+
+    # userAccountControl is a bitmask used to store a number of settings.
+    # find more at https://support.microsoft.com/en-gb/help/305144/how-to-use-the-useraccountcontrol-flags-to-manipulate-user-account-pro
+
+    COOMON_ACCOUNT_CONTROL_FLAGS = {
+        512: "Enabled Account",
+        514: "Disabled account",
+        544: "Account Enabled - Require user to change password at first logon",
+        4096: "Workstation/server",
+        66048: "Enabled, password never expires",
+        66050: "Disabled, password never expires",
+        66080: "Enables, password never expires, password not required.",
+        532480: "Domain controller"
+    }
+    NORMAL_ACCOUNT = 512
+    DISABLED_ACCOUNT = 514
+
+    # common attributes for specific AD objects
+    DEFAULT_PERSON_ATTRIBUTES = [
+        'name',
+        'displayName',
+        'memberOf',
+        'mail',
+        'samAccountName',
+        'manager',
+        'userAccountControl'
+    ]
+    DEFAULT_COMPUTER_ATTRIBUTES = [
+        'name',
+        'memberOf'
+    ]
+
+    ''' HELPER FUNCTIONS '''
+
+    def error_entry(message):
+        return {
+            'Type': entryTypes['error'],
+            'Contents': message,
+            'ContentsFormat': formats['text']
+        }
+
+    def initialize_server():
+        """
+        uses the instance configuration to initialize the LDAP server
+        """
+        if SECURE_CONNECTION == "SSL":
```

Comment:
Note that in current implementation ssl and tls don't refer to protocols. The naming that was chosen is not very clear. ssl is referring to an ldap server that listens on an ssl port (usually 636). tls refers to an ldap server listening on a plain tcp socket but supporting starttls (see:  https://ldap3.readthedocs.io/bind.html?highlight=starttls#starttls). Additionally, there should be the "insecure" option.

Reply:
We had a discussion about this issue, since in the previous integration 'port', 'connection type' and 'certificate validation' are all configurable, but have dependencies between them.

[here](https://ldap3.readthedocs.io/ssltls.html) initializing the server object with a TLS object is used for certificate validation (the TLS object itself needs to be initialized with correct parameters. needs more thinking..) while start_tls() is used to start a secure connection over clear-text (usually port 389).